### PR TITLE
[EC] Unify scalar_mul_base point for ec_nistp curves

### PR DESCRIPTION
**Pull request by dkostic**
### Issues:
CryptoAlg-2544

### Description of changes: 
Added unified scalar multiplication of the base point for curves
implemented in ec_nistp. This is a refactor of the algorithm in
p384.c and p521.c that makes it generic. The implementation 
in p384.c and p521.c is substituted with this new unified
implementation.

### Call-outs:
Point out areas that need special attention or support during the review process. Discuss architecture or design changes.

### Testing:
How is this change tested (unit tests, fuzz tests, etc.)? Are there any testing steps to be verified by the reviewer?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 1]
```diff
@@ -498,3 +498,142 @@ void ec_nistp_scalar_mul(const ec_nistp_meth *ctx,
...
+  // Process the 4 groups of digits starting from group (3) down to group (0).
+  for (int i = 3; i >= 0; i--) {
+    // Double |res| 5 times in each iteration, except in the first one.
+    for (size_t j = 0; i != 3 && j < SCALAR_MUL_WINDOW_SIZE; j++) {
+      ctx->point_dbl(x_res, y_res, z_res, x_res, y_res, z_res);
+    }
+
+    // Process the digits in the current group from the most to the least
+    // significant one (this is a requirement to ensure that the case of point
+    // doubling can't happen).
```
[the remark this answers is not in the document]
Comment (dkostic) [reply crypto/fipsmodule/ec/ec_nistp.c:593]: yes, makes sense, I'll update

File: crypto/fipsmodule/ec/p521.c [hunk 2]
```diff
@@ -433,18 +429,6 @@ OPENSSL_STATIC_ASSERT(P521_MUL_WSIZE == 5,
 
 // p521_select_point_affine selects the |idx|-th affine point from
 // the given precomputed table and copies it to |out| in constant-time.
```
Comment (samuel40791765) [inline crypto/fipsmodule/ec/p521.c:431]: Should this comment also be removed?

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 3]
```diff
@@ -498,3 +498,141 @@ void ec_nistp_scalar_mul(const ec_nistp_meth *ctx,
...
+
+  // Process the 4 groups of digits starting from group (3) down to group (0).
+  for (int i = 3; i >= 0; i--) {
+    // Double |res| 5 times in each iteration, except in the first one.
+    for (size_t j = 0; i != 3 && j < SCALAR_MUL_WINDOW_SIZE; j++) {
+      ctx->point_dbl(x_res, y_res, z_res, x_res, y_res, z_res);
+    }
+
+    // Process the digits in the current group from the most to the least
+    // significant one.
```
Comment (samuel40791765) [inline crypto/fipsmodule/ec/ec_nistp.c:592]: Should we keep the original relevant comment? It does seem a bit important, but I'm not 100% sure.
```suggestion
    // Process the digits in the current group from the most to the least
    // significant one (this is a requirement to ensure that the case of point
    // doubling can't happen).
```

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 4]
```diff
@@ -498,3 +498,141 @@ void ec_nistp_scalar_mul(const ec_nistp_meth *ctx,
...
+
+      // Negate y coordinate of the point tmp = (x, y); ftmp = -y.
+      ec_nistp_felem ftmp;
+      ctx->felem_neg(ftmp, y_tmp);
+
+      cmovznz(y_tmp, ctx->felem_num_limbs, is_neg, y_tmp, ftmp);
+
+      // Add the point to the accumulator |res|.
+      ctx->point_add(x_res, y_res, z_res, x_res, y_res, z_res, 1,
+                     x_tmp, y_tmp, ctx->felem_one);
```
Comment (samuel40791765) [inline crypto/fipsmodule/ec/ec_nistp.c:620]: Should we retain the documentation around `p384/521_point_add`?

File: crypto/fipsmodule/ec/ec_nistp.c (hunk 3 above)
Comment (dkostic) [reply crypto/fipsmodule/ec/ec_nistp.c:592]: https://github.com/aws/aws-lc/pull/2003#discussion_r1852681259

File: crypto/fipsmodule/ec/p521.c (hunk 2 above)
Comment (dkostic) [reply crypto/fipsmodule/ec/p521.c:431]: yes, I removed it in the next PR: https://github.com/aws/aws-lc/pull/2004

File: crypto/fipsmodule/ec/ec_nistp.c (hunk 4 above)
Comment (dkostic) [reply crypto/fipsmodule/ec/ec_nistp.c:620]: not needed imo, but if you insist I'll add it (maybe in the next pr if no other changes are needed for this one).

**Pull request merged**